File: articles/hash-functions-integrity-and-verification/typescript/hash_verification_demo.ts
// TypeScript hash verification teaching demo using Node's crypto module.
import { createHash } from "crypto";

function sha256Prefix(input: string): string {
  return createHash("sha256").update(input, "utf8").digest("hex").slice(0, 16);
}

const original = "verified artifact manifest";
const altered = "verified artifact manifest!";
console.log(`original sha256=${sha256Prefix(original)}`);
console.log(`altered sha256=${sha256Prefix(altered)}`);
console.log(`match=${sha256Prefix(original) === sha256Prefix(altered)}`);

File: articles/hash-functions-integrity-and-verification/java/HashVerificationDemo.java
import java.security.MessageDigest;
import java.nio.charset.StandardCharsets;

public class HashVerificationDemo {
    static String sha256Prefix(String input) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] bytes = digest.digest(input.getBytes(StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.substring(0, 16);
    }

    public static void main(String[] args) throws Exception {
        String original = "verified artifact manifest";
        String altered = "verified artifact manifest!";
        System.out.println("original sha256=" + sha256Prefix(original));
        System.out.println("altered sha256=" + sha256Prefix(altered));
        System.out.println("match=" + sha256Prefix(original).equals(sha256Prefix(altered)));
    }
}
